File: 0688-knight-probability-in-chessboard/solution.py
class Solution:
    
    def knightProbability(self, n: int, k: int, row: int, col: int) -> float:
        # get all steps for a knight
        steps = [(1,2),(-1,2),(1,-2),(-1,-2),(2,1),(-2,1),(-2,-1),(2,-1)]
        
        # Square represents the probability that a knight is on the chess board
        probSquare           = [[0]*n for _ in range(n)]
        
        # Since the knight is initially on (row, col) the probavility is 1
        probSquare[row][col] = 1
                
        # Iterate for k moves
        for move in range(1, k+1):
            # initialize new prob square
            newProbSquare = [[0]*n for _ in range(n)]
            
        #for each position in chessboard sum the probaility to get to the position
        #since each move is uniformly and randmly made the probability should 
        #be muliplied by 1/8 (0.125)
        
#prob(chess[k] [x][y] = sum(1/8 * prob(chess[k-1] [x+dx][y+dy]))) for each step   
            
            for x in range(n):
                for y in range(n):
                    for step in steps:
                        dx, dy = step
                        if 0 <= (x+dx) < n and 0 <= (y+dy) < n:
                            newProbSquare[x][y] += (0.125*probSquare[x+dx][y+dy])
            
            probSquare = newProbSquare

        # At the end we sum the probailities in the chess board
        ans = sum(sum(row) for row in probSquare)
                
        return ans

File: 0688-knight-probability-in-chessboard/solution.cpp
class Solution {
public:
    double knightProbability(int n, int k, int row, int col) 
    {
        const int numSteps = 8;
        
        const double P     = 0.125;
        
        const int steps[numSteps][2] = {{1,2},{-1,2},{1,-2},{-1,-2},{2,1},{-2,1},{-2,-1},{2,-1}};
    
        vector<vector<double>> probSquare(n, vector<double>(n, 0.0));
        probSquare[row][col] = 1;
        
        
        for(int move = 1; move <= k; move++)
        {
            vector<vector<double>> newProbSquare(n, vector<double>(n, 0.0));
            for(int x = 0; x < n; x++)
            {
                for(int y = 0; y < n; y++)
                {
                    for(int step = 0; step < numSteps; step++)
                    {
                        int dx, dy; dx = steps[step][0]; dy = steps[step][1];
                        if((0 <= x+dx && x+dx < n) && (0 <= y+dy && y+dy < n))
                        {
                            newProbSquare[x][y] += P*probSquare[x+dx][y+dy];
                        }
                    }
                }
            }
            
            probSquare = newProbSquare;
        }
        
        double ans = 0.0;
        for(int row = 0; row < n; row++)
        {
            for(int col = 0; col < n; col++)
            {
                ans += probSquare[row][col];
            }
        }
        
        return ans;
    
    }
};
